### DISCUSSION TITLE
Exercise 25: Consider an asynchronous system of 5 processes that run the Ben-Or

### DISCUSSION THREAD

**Discussion by rotiroti** [sapienzastudentsnetwork/distributed-systems#50] [Exercises]
Consider an asynchronous system of 5 processes that run the Ben-Or randomized consensus algorithm. The number of failures that the system allows is 2. Show that, if at 2 failures occurs, then the probability that the protocol terminates after $x$ rounds is smaller than $\alpha^x$, for some $\alpha$.

**Comment by anto707** [comment]
Could this be modelled with a geometric distribution? If we see the probability of success (termination in a specific round) as the probability of picking all the same bits (all 0s or all 1s) in the coin flip step?

**Comment by riccardoscuto** [comment]
I asked BingAI and he wrote this, what do you think?

The Ben-Or randomized consensus algorithm is a fundamental algorithm in distributed computing, designed to solve consensus in an asynchronous message-passing system where processes can fail by crashing.

In the specific case, where there are 5 processes and up to 2 can fail, the algorithm can still reach consensus if a majority of processes are correct. However, the termination of the algorithm is probabilistic, and it’s not guaranteed to terminate in a fixed number of rounds.

The probability of termination after $x$ rounds being smaller than $\alpha^x$, for some $\alpha$, is a common way to express the fact that the probability of termination decreases exponentially with the number of rounds. This is due to the randomized nature of the algorithm: in each round, each process flips a coin (i.e., generates a random bit), and the decision (consensus) is based on the majority of the coin flips. If there is no majority (i.e., a tie), the processes proceed to the next round and repeat the process.

**Comment by aleiis** [comment]
The variable $X \sim \text{Geom}(p)$ represents the number of rounds required for the algorithm to achieve a round success with a probability of $p$. In order for a round to be considered successful, a minimum of three processes must have the same preference, either '1' or '0'.

The probability $p$ depends on the number of failures, and there are three distinct cases that determine its value:
- If $F = 0$ then $p = 1$
- if $F = 1$ then $p = \frac{10}{16}$
- If $F = 2$ then $p = \frac{2}{8}$

Since $X$ follows a geometric distribution, the probability that the protocol terminates after $x$ rounds (or more) is:
$P(X > x) = (1-p)^x$.

**Reply by notherealmarco** [reply]
To complete your answer, for F = 2 we obtain p = 1/4. This is the smallest value among the three possible choices for F. So we can say that p >= 1/4.

With some simple algebra:
p >= 1/4 ==> -p <= -1/4 ==> 1-p <= 1/4

So given that 1-p <= 1/4, we can say that P(X > x) <= (1 - 1/4)^x = (3/4)^x

Finally, the probability that the protocol terminates after x rounds is smaller than α^x for any α > 3/4.

**Reply by cornflexxx** [reply]
I think it's possible to get a weaker lower bound on $p$ by doing less calculations. In fact we know that 

$$p = \Pr[\text{Success round}]=  \Pr[\text{All preferences are the same}] + P= 2/32 + P $$

Assuming the worst case,  $f = 2$ we get that $p \ge 2/32$, therefore $1-p \le 30/32$. As you said we can model r.v. $X \sim Geom(p)$ and get 

$$\Pr[X>x] = \Pr[\text{protocol terminates after $x$ rounds}] \le (1-p)^x \le \frac{30}{32}^x$$

**Comment by simonesestito** [comment]
> The probability $p$ depends on the number of failures, and there are three distinct cases that determine its value:
> 
> * If $F = 0$ then $p = 1$

How can $p$ be equal to $1$?

**Reply by notherealmarco** [reply]
for F = 0 we need 3 nodes out of 5 to randomly pick the same preference.
That is P(at least 3 nodes pick 0) = P(exactly 3 pick 0) + P(exactly 4 pick 0) + P(exactly 5 pick 0).

To compute P(exactly k nodes pick 0), we can model a binomial distribution Y ∼ Binomial(n, p) where n = 5 (all nodes are alive if F=0!) and p = 1/2 (the probability of a single node obtaining 0).

Now we can compute P(Y = 3) = (5 choose 3) * (1/2)^3 * (1/2)^2 = 10/32. And in this way we also obtain that P(Y = 4) = 5/32 and P(Y = 5) = 1/32.

Now to compute P(at least 3 nodes pick 0) we just need to sum them and we obtain 16/32 = 1/2.

Of course we don't care if they agree on 0 or 1, we only care that they agree on the same value. So P(at least 3 nodes agree on a value) = P(at least 3 nodes pick 0) + P(at least 3 nodes pick 1) = 2*(1/2) = 1

**Comment by aleiis** [comment]
> 
> How can $p$ be equal to $1$?

If there are no failures the system with 5 participants will always reach a majority either on '0' or '1' on the first round.
